### PR TITLE
[SC] Cache Mining SC Results for validation

### PR DESCRIPTION
**Pull request by codingupastorm**

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.SmartContracts/Rules/SmartContractCoinViewRuleLogic.cs [hunk 1]
```diff
@@ -102,6 +126,14 @@ public async Task RunAsync(Func<RuleContext, Task> baseRunAsync, RuleContext con
         /// </summary>
         public void UpdateCoinView(Action<RuleContext, Transaction> baseUpdateUTXOSet, RuleContext context, Transaction transaction)
         {
+            // We already have results for this block. No need to do any processing other than updating the UTXO set.
+            if (this.cachedResults != null)
+            {
```
Comment (rowandh) [inline src/Stratis.Bitcoin.Features.SmartContracts/Rules/SmartContractCoinViewRuleLogic.cs]: This bypasses `ValidateRefunds`, `ValidateGeneratedTransaction` and `ValidateSubmittedTransaction`.

File: src/Stratis.Bitcoin.Features.SmartContracts/Rules/SmartContractCoinViewRuleLogic.cs [hunk 2]
```diff
@@ -74,9 +79,28 @@ public async Task RunAsync(Func<RuleContext, Task> baseRunAsync, RuleContext con
             uint256 blockRoot = ((ISmartContractBlockHeader)context.ValidationContext.ChainedHeaderToValidate.Previous.Header).HashStateRoot;
 
             this.logger.LogDebug("Block hash state root '{0}'.", blockRoot);
-            byte[] blockRootBytes = blockRoot.ToBytes();
-            this.mutableStateRepository = this.stateRepositoryRoot.GetSnapshotTo(blockRootBytes);
 
+            this.cachedResults = this.executionCache.GetExecutionResult(block.GetHash());
```
Comment (rowandh) [inline src/Stratis.Bitcoin.Features.SmartContracts/Rules/SmartContractCoinViewRuleLogic.cs]: Not an issue with correctness, but to improve readability could we set this back to `null` after it's finished being used?

Comment (codingupastorm) [reply src/Stratis.Bitcoin.Features.SmartContracts/Rules/SmartContractCoinViewRuleLogic.cs]: I think to keep it in-line with the other variables, like `refundCounter` it makes a lot of sense for it to be set only here. Each new run has the responsibility of setting these correctly initially.

File: src/Stratis.Bitcoin.Features.SmartContracts/Rules/SmartContractCoinViewRuleLogic.cs (hunk 1 above)
Comment (codingupastorm) [reply src/Stratis.Bitcoin.Features.SmartContracts/Rules/SmartContractCoinViewRuleLogic.cs]: Right, I'm willing to say that it's okay as we are trusting our own miner, it's part of the same program.

**Pull request merged**